janusgraph-core/src/main/java/org/janusgraph/diskstorage/util/BufferUtil.java

```diff
@@ -57,6 +65,51 @@ public static StaticBuffer getIntBuffer(int[] ids) {
         return StaticArrayBuffer.of(arr);
     }
 
+    /**
+     * We always add a byte marker at the beginning to indicate it's a string buffer. Then we
+     * write the string uncompressed. Finally, if the total buffer size is 8, we add a dummy byte
+     * at the end so that upon read time, JanusGraph knows the buffer does not store a long value.
+     * See {@link IDManager#getKeyID(StaticBuffer)} for more details.
+     *
+     * @param s
+     * @return
+     */
+    public static StaticBuffer getStringBuffer(String s) {
+        if (StringUtils.isEmpty(s) || !StringEncoding.isAsciiString(s)) {
+            throw new IllegalArgumentException("value must be non-empty ASCII string but received: " + s);
+        }
+        ByteBuffer buffer;
+        if (s.length() + byteSize == longSize) {
+            buffer = ByteBuffer.allocate(longSize + byteSize);
+        } else {
+            buffer = ByteBuffer.allocate(s.length() + byteSize);
+        }
+        buffer.put(STRING_ID_MARKER);
+        for (int i = 0; i < s.length(); i++) {
+            int c = s.charAt(i);
+            assert c <= 127;
+            byte b = (byte)c;
+            /**
+             * we don't have to apply STOP_MASK here because a static buffer
+             * has a fixed length, and thus upon read time, we know where
+             * to stop reading the string. This is more to keep it consistent
+             * with {@link StringEncoding#writeAsciiString(byte[], int, String)}
+             * where we use STOP_MASK to mark the end of the string.
+             *
+             * An additional benefit of doing so is to enable corruption check
+             * in the future. Note: this approach has no overhead.
+             */
+            if (i+1==s.length()) b |= STOP_MASK;
+            buffer.put(b);
+        }
+        if (s.length() + byteSize == longSize) {
+            // this could be any dummy byte
+            buffer.put(STOP_MASK);
+        }
```

Comment:
This means Strings with size == 7 and Strings with size == 8 both will be converted to the String with size == 9.
I told about that 4 bits of `STRING_ID_MARKER` could be used as a data type. I now think 1 bit from that could also be used to tell if the encoded size of the value if less then 8 bits or not. If yes, then we know that the string is of size 7. Otherwise it's 8.
It will help to use UTF-8 characters here. This is can also be applicable for `byte[]` data type in the future.
Thus the first byte marker could have 3 reserved bits, 1 bit for value size detection (less then 8 bits or not), and 4 bits for data type detection (String, byte[], UUID, etc.).

Reply:
> I now think 1 bit from that could also be used to tell if the encoded size of the value if less then 8 bits or not. If yes, then we know that the string is of size 7. Otherwise it's 8.

Make sense, perfect.